libutils/pcre_wrap.c

```diff
@@ -0,0 +1,495 @@
+{
+    if (error < 0)
+    {
+        switch (error)
+        {
+            /* Passed-through PCRE error: */
+        case PCRE_ERROR_NOMEMORY:     return "(pcre:) No memory";
+
+            /* Shouldn't happen unless PCRE or PCRE_WRAP bug, or user messed with compiled job: */
+        case PCRE_ERROR_NULL:         return "(pcre:) NULL code or subject or ovector";
+        case PCRE_ERROR_BADOPTION:    return "(pcre:) Unrecognized option bit";
+        case PCRE_ERROR_BADMAGIC:     return "(pcre:) Bad magic number in code";
+        case PCRE_ERROR_UNKNOWN_NODE: return "(pcre:) Bad node in pattern";
+
+            /* Can't happen / not passed: */
+        case PCRE_ERROR_NOSUBSTRING:  return "(pcre:) Fire in power supply"; 
+        case PCRE_ERROR_NOMATCH:      return "(pcre:) Water in power supply";
+
+            /* PCRE_WRAP errors: */
+        case PCRE_WRAP_ERR_NOMEM:          return "(pcre_wrap:) No memory";
+        case PCRE_WRAP_ERR_CMDSYNTAX:      return "(pcre_wrap:) Syntax error while parsing command";
+        case PCRE_WRAP_ERR_STUDY:          return "(pcre_wrap:) PCRE error while studying the pattern";
+        case PCRE_WRAP_ERR_BADJOB:         return "(pcre_wrap:) Bad job - NULL job, pattern or substitute";
+        case PCRE_WRAP_WARN_BADREF:        return "(pcre_wrap:) Backreference out of range";
+
+            /* What's that? */
+        default:  return "Unknown error";
+        }
+    }
+    /* error >= 0: No error */
+    return "(pcre_wrap:) Everything's just fine. Thanks for asking.";
+
+}
+
+static int pcre_wrap_parse_perl_options(const char *options, int *flags)
+{
+    int rc = 0;
+    *flags = 0;
+
+    if (NULL == options)
+    {
+        return 0;
+    }
+
+    size_t max = strlen(options);
+    for (size_t i = 0; i < max; i++)
+    {
+        switch(options[i])
+        {
+        case 'e': break; /* Nein! Jamais! Nunca! Well, maybe if you ask nicely... */
+        case 'g': *flags |= PCRE_WRAP_GLOBAL; break;
+        case 'i': rc |= PCRE_CASELESS; break;
+        case 'm': rc |= PCRE_MULTILINE; break;
+        case 'o': break;
+        case 's': rc |= PCRE_DOTALL; break;
+        case 'x': rc |= PCRE_EXTENDED; break;
+        case 'U': rc |= PCRE_UNGREEDY; break;
+        case 'T': *flags |= PCRE_WRAP_TRIVIAL; break;
+        default: break;
+        }
+    }
+    return rc;
+
+}
+
+static pcre_wrap_substitute *pcre_wrap_compile_replacement(const char *replacement, int trivialflag, int capturecount, int *errptr)
+{
+    int refnum = 0;
+    char *text;
+    pcre_wrap_substitute *rstruct;
+
+    /*
+     * Sanity check
+     */
+    if (NULL == replacement)
+    {
+        replacement = "";
+    }
+    int length = (int)strlen(replacement); // intentional
+
+    rstruct = (pcre_wrap_substitute *)calloc(1, sizeof(pcre_wrap_substitute));
+    if (NULL == rstruct)
+    {
+        *errptr = PCRE_WRAP_ERR_NOMEM;
+        return NULL;
+    }
+
+    text = (char *)calloc(1, length + 1);
```

Comment:
It's not a style issue, [this cast is dangerous](http://c-faq.com/malloc/mallocnocast.html). Its roots are in the pre-ANSI C era, but today it has a negative effect. So yes, please do a quick search for "alloc" and remove the casts.